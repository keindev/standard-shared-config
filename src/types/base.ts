export enum FileType {
  Text = 'text',
  GLOB = 'glob',
  JSON = 'json',
  YAML = 'yaml',
}

export enum EntityName {
  Scripts = 'scripts',
  Snapshots = 'snapshots',
  Package = 'package',
}

export type IMergeRule = string[] | null | boolean;
export type IScript = [string, string];
export type IMergeMap = { [key: string]: IMergeRule };

export interface ISnapshot {
  content: string;
  executable: boolean;
  hash: string;
  merge: IMergeRule;
  path: string;
  type: FileType;
}
