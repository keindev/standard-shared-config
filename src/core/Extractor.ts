import Package from 'package-json-helper';
import { PackageManager } from 'package-json-helper/lib/types';
import { cast } from 'package-json-helper/lib/utils/parsers';
import path from 'path';
import TaskTree from 'tasktree-cli';
import yaml from 'yaml';

import {
    CONFIG_FILE, EntityName, FileType, IDependency, IExtractionConfig, IExtractionOptions, INormalizedExtractionConfig,
    ISnapshot,
} from '../types.js';
import { getHash, mergeFiles, readFile, writeFile } from '../utils/file.js';

export default class Extractor {
  #sharedDir: string;

  constructor(sharedDir: string) {
    this.#sharedDir = sharedDir;
  }

  async extract({ snapshots = [], ...options }: IExtractionOptions): Promise<void> {
    const config = await this.readConfig(options);
    const pkg = new Package(path.join(process.cwd(), 'package.json'), config.package.manager ?? PackageManager.NPM);

    await pkg.read();
    await this.extractFiles(snapshots);
    this.updatePackage(config, pkg);
    await pkg.save();
    await this.installDependencies(config.dependencies, pkg);
  }

  private async extractFiles(snapshots: ISnapshot[]): Promise<void> {
    const task = TaskTree.add('Processing config files...');
    const extract = async (snapshot: ISnapshot): Promise<void> => {
      const filePath = path.resolve(process.cwd(), snapshot.path);
      const extendFilePath =
        snapshot.type === FileType.GLOB ? snapshot.path : path.join(this.#sharedDir, snapshot.path);
      const extendFileData = snapshot.merge ? await readFile(path.resolve(process.cwd(), extendFilePath)) : false;

      if (extendFileData) {
        await writeFile(filePath, mergeFiles(snapshot, extendFileData), snapshot.executable);
        task.log(`${snapshot.path} merged`);
      } else {
        const currentFileData = await readFile(filePath);
        const hash = currentFileData ? getHash(currentFileData) : undefined;

        if (hash !== snapshot.hash) {
          await writeFile(filePath, snapshot.content, snapshot.executable);
          task.log(`${snapshot.path} ${currentFileData ? 'updated' : 'created'}`);
        }
      }
    };

    await Promise.all(snapshots.map(extract));
    task.complete('Processed configs:');
  }

  private async installDependencies(dependencies: IDependency[], pkg: Package): Promise<void> {
    const task = TaskTree.add('Check package devDependencies:');
    const installationMap = new Map<string, string | undefined>();

    dependencies.forEach(([name, version]) => {
      const dependency = pkg.devDependencies.get(name);

      if (!dependency) installationMap.set(name, version);
      else if (version && !dependency.isSatisfy(version)) task.error(`${name} version is must be: ${version}`, false);
    });

    if (task.haveErrors) task.fail('Some dependencies are missing or have wrong version:');
    if (installationMap.size) {
      const subtask = task.add('Installing packages:');

      installationMap.forEach((version, name) => subtask.log(`{bold ${name}}: ${version ?? '{italic latest}'}`));
      await pkg.install(installationMap, ['--save-dev']);
      subtask.complete('Installed packages');
    }

    task.complete('All dependencies from shared config is presented in package.json');
  }

  private async readConfig({
    dependencies = [],
    scripts = [],
    ...options
  }: Omit<IExtractionOptions, EntityName.Snapshots>): Promise<INormalizedExtractionConfig> {
    const content = await readFile(path.resolve(process.cwd(), this.#sharedDir, CONFIG_FILE));
    const config: IExtractionConfig = content ? yaml.parse(content) : {};
    const overrideScripts = new Map(Object.entries(config.overrideScripts ?? {}));
    const ignoreDependencies = new Set(config.ignoreDependencies ?? []);

    return {
      scripts: scripts.map(([key, value]) => [key, overrideScripts.get(key) ?? value]),
      dependencies: dependencies.filter(([key]) => !ignoreDependencies.has(key)),
      package: {
        exports: cast.toExportsMap(options.package?.exports),
        manager: options.package?.manager ?? PackageManager.NPM,
        peerDependencies: options.package?.peerDependencies,
        type: options.package?.type,
        types: cast.toString(options.package?.types),
      },
    };
  }

  private updatePackage(
    { scripts, package: { type, types, exports, peerDependencies = {} } }: INormalizedExtractionConfig,
    pkg: Package
  ): void {
    let task = TaskTree.add('Inserting the required scripts:');

    scripts.forEach(([key, value]) => {
      const script = pkg.scripts.get(key);

      if (!script) task.log(`{bold ${key}} script added`);
      if (script && script !== value) task.log(`{bold ${key}} script replaced`);

      pkg.scripts.set(key, value);
    });

    cast.toDependencyMap(peerDependencies, pkg.peerDependencies);
    task.complete('Package scripts was updated!');
    task = TaskTree.add('Update package params:');

    if (type && pkg.type !== type) pkg.type = type;
    if (types && pkg.types !== types) pkg.types = types;
    if (exports) {
      if (pkg.exports) {
        const { map } = pkg.exports;

        exports.map.forEach((value, key) => map.set(key, value));
      } else {
        pkg.exports = exports;
      }
    }

    task.complete('Package parameters was updated!');
  }
}
