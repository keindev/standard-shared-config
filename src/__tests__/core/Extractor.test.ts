// @see https://github.com/facebook/jest/issues/9430
import { promises as fs } from 'fs';
import { glob } from 'glob';
import path from 'path';

// eslint-disable-next-line node/no-extraneous-import
import { jest } from '@jest/globals';

import Extractor from '../../core/Extractor.js';
import { FileType } from '../../types/base.js';

jest.useFakeTimers();

jest.spyOn(glob, 'sync').mockImplementation(() => ['config1.json', 'test/config2.json']);
jest.spyOn(fs, 'access').mockImplementation(() => Promise.resolve());
jest.spyOn(fs, 'unlink').mockImplementation(() => Promise.resolve());
jest.spyOn(fs, 'mkdir').mockImplementation(() => Promise.resolve(''));
jest.spyOn(fs, 'readFile').mockImplementation(filePath => {
  const basename = path.basename(filePath as string);
  let content = '';

  if (basename === '.gitignore') content = 'logs\ncoverage\n*.log\nyarn-debug.log*\nyarn-error.log*\n.env';
  if (basename === 'package.json') {
    content = JSON.stringify({ name: 'test', devDependencies: { '@types/jest': '^1.x', 'ts-jest': '^26.x' } });
  }

  return Promise.resolve(content);
});

describe('Extractor', () => {
  const extractor = new Extractor('.config');
  const files: [string, string][] = [];
  const appendFile = (name: any, data: any): void => {
    files.push([path.relative(process.cwd(), name.toString()), data.toString()]);
  };

  jest.spyOn(fs, 'writeFile').mockImplementation((name, data) => Promise.resolve(appendFile(name, data)));

  it('extract', async () => {
    await extractor.extract({
      scripts: [
        ['test', 'jest'],
        ['build', 'tsc'],
      ],
      snapshots: [
        {
          path: 'test/config2.json',
          hash: '1',
          merge: false,
          executable: false,
          type: FileType.JSON,
          content: JSON.stringify({}),
        },
        {
          path: 'config1.json',
          hash: '2',
          merge: false,
          executable: false,
          type: FileType.JSON,
          content: JSON.stringify({}),
        },
        {
          path: '.gitignore',
          hash: '3',
          merge: true,
          executable: false,
          type: FileType.GLOB,
          content: '.env\nnode_modules/',
        },
      ],
      package: {
        peerDependencies: {
          react: '18.x',
          'react-dom': '18.x',
        },
      },
    });

    expect(files).toMatchSnapshot();
  });
});
