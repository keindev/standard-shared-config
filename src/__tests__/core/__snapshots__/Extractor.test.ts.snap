// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extractor extract 1`] = `
Array [
  Array [
    "test/config2.json",
    "{}",
  ],
  Array [
    "config1.json",
    "{}",
  ],
  Array [
    ".gitignore",
    "*.log
.env
coverage
logs
node_modules/
yarn-debug.log*
yarn-error.log*",
  ],
  Array [
    "package.json",
    "{
  \\"name\\": \\"test\\",
  \\"type\\": \\"commonjs\\",
  \\"scripts\\": {
    \\"build\\": \\"tsc\\",
    \\"test\\": \\"jest\\"
  },
  \\"devDependencies\\": {
    \\"@types/jest\\": \\"^1.x\\",
    \\"ts-jest\\": \\"^26.x\\"
  }
}
",
  ],
  Array [
    "package.json",
    "[[\\"jest\\",\\"8.x\\"]]",
  ],
]
`;
